sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java

```diff
@@ -173,15 +173,15 @@ public boolean start() {
 
     @Override
     public boolean advance() {
-      if (current >= numMessagesPerShard) {
+      if (current >= numMessagesPerShard - 1) {
         return false;
       }
```

Comment:
I prefer this order - if done early exit, if want to repeat repeat, otherwise advance to valid position.